### PR TITLE
Merge `unused_tuple_struct_fields` into `dead_code`

### PR DESCRIPTION
**Pull request by shepmaster**
This implicitly upgrades the lint from `allow` to `warn` and places it into the `unused` lint group.

[Discussion on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Moving.20.60unused_tuple_struct_fields.60.20from.20allow.20to.20warn)

### CODE REVIEW

**Comment by compiler-errors** [comment]
I'd happily take a PR that proactively fixes all of the compiler warnings. Most of them can just be removed, except for the `TyCtxt` one, which can be replaced with `PhantomData`.

**Comment by shepmaster** [comment]
Yeah, my plan is to find all of the current issues and submit it as a separate PR, it’s just easier to find it here first.

I was attempting to do that without causing undue noise (thanks rustbot 👀).

File: compiler/rustc_infer/src/infer/nll_relate/mod.rs
Comment (Nadrieril) [inline compiler/rustc_infer/src/infer/nll_relate/mod.rs:117]: That one is getting nuked by https://github.com/rust-lang/rust/pull/118257, you may as well remove the whole struct.

Comment (shepmaster) [reply compiler/rustc_infer/src/infer/nll_relate/mod.rs:117]: Sounds good; I'll wait until that is merged to avoid rework / conflicts.

**Comment by traviscross** [comment]
Regarding the T-lang nomination, one of the open questions here is whether this should be part of the `dead_code` lint now that we're making it warn-by-default.

The motivation for making this part of `dead_code` is that unused fields of normal structs warn under `dead_code`, so this would be consistent.  E.g.:

```rust
warning: field `a` is never read
 --> src/lib.rs:1:12
  |
1 | struct A { a: () }
  |        -   ^
  |        |
  |        field in this struct
  |
  = note: `#[warn(dead_code)]` on by default
```

On 2022-01-25, T-lang [requested](https://github.com/rust-lang/rust/pull/92972#issuecomment-1021511970) this lint be separated from `dead_code` so that it could be introduced gradually and so people could disable it specifically.  The motivation seems to have been the belief that this lint could be "harder to fix than removing named fields, because removing a positional field changes the indices of the following fields".

However, after that decision, on 2022-01-26, @pnkfelix [suggested](https://github.com/rust-lang/rust/pull/92972#issuecomment-1021815408) a change to the lint that would allow people to resolve the warning without changing field indices:

> ...I thought: "well, we should let people include () fields (i.e. not warn about them), and perhaps even have the lint diagnostic here suggest either removing the field or making it have () type. That would keep the indices unchanged."

This change was [implemented](https://github.com/rust-lang/rust/pull/92972#issuecomment-1096739201) and is suggested as part of the error message, e.g.:

```rust
error: field is never read: `1`
  --> $DIR/tuple-struct-field.rs:6:21
   |
LL | struct Wrapper(i32, [u8; LEN], String);
   |                     ^^^^^^^^^
   |
help: change the field to unit type to suppress this warning while preserving the field numbering
   |
LL | struct Wrapper(i32, (), String);
   | 
```

In discussion in T-compiler in Zulip, the feeling seemed to be that this should be part of `dead_code`.  Given the change suggested by @pnkfelix and implemented in the lint that resolves the issue of preserving field indices, do we now agree?

**Comment by shepmaster** [comment]
> a PR that proactively fixes all of the compiler warnings.

Opened in #118382

**Comment by shepmaster** [comment]
r? compiler

File: compiler/rustc_passes/src/dead.rs
Comment (WaffleLapkin) [inline compiler/rustc_passes/src/dead.rs:1017]: Isn't this just `is_positional = is_pos` or `is_positional |= is_pos`?

Comment (WaffleLapkin) [reply compiler/rustc_passes/src/dead.rs:1017]: Also can't we just do a single check is variant is tuple-like or not, instead of trying to move this out of the loop?..

**Review by WaffleLapkin — CHANGES_REQUESTED** [review]
Just a bit of a nitpick. Feel free to change labels to waiting on review once CI is green

**Comment by shepmaster** [comment]
There are a few disparate commits / change types:

- Removing actually unused tuple fields
  - From Clippy
  - From Rustfmt
- The core lint change
- Renaming all uses of `unused_tuple_struct_fields`
  - In the primary test suite
  - In the Clippy test suite
- New `allow(...)` where fields started causing warnings / test failures
  - In the primary test suite
  - In the Miri test suite

Do the Clippy and Rustfmt teams want the unused tuple field removal commits as separate PRs? If so, to which repo?

What are the feelings on the rest of the changes though? They are kinda of chunky and feel likely to bitrot.

File: compiler/rustc_passes/src/dead.rs
Comment (shepmaster) [reply compiler/rustc_passes/src/dead.rs:1017]: > Isn't this [...] `is_positional |= is_pos`?

Yeah, I thought about that but wanted to make the diff smaller / easier to see how it changed. Happy to update to this.

> do a single check is variant is tuple-like or not

That seems plausible, but I don't know my way around the compiler enough to do so. Got any hints?

> instead of trying to move this out of the loop

I thought you were suggesting that we *should* move it out of the loop — I must not understand what you mean.

**Comment by calebcartwright** [comment]
> Do the Clippy and Rustfmt teams want the unused tuple field removal commits as separate PRs? If so, to which repo?

Speaking only for the rustfmt side, but yes probably.

In general I dislike non-mandatory changes to the subtrees being made here because I've had one too many instances where small, innocuous changes made in this repo end up making my life disproportionately more difficult on subtree syncs :sweat_smile: 

What I'm still a little unclear on is whether non-build failing lints are run against subtree codebases in this repo, and perhaps more importantly, whether they should be.

More to the point, if this is something that's moving towards being a build breaking lint in this repo, then my preference would be to update the code in r-l/rustfmt, albeit with the dead code permitted; we do have plans for the currently unused span field, and having the field there will help me remember what I was thinking whenever I get back around to it.

**Comment by shepmaster** [comment]
> More to the point, if this is something that's moving towards being a build breaking lint in this repo

There's two steps happening at once here:

1. The lint known as `unused_tuple_struct_fields` is moving from allow- to warn-by-default. This would cause the build to break by itself.
2. The lint is being merged into the existing `dead_code` lint. This makes it harder to phase-in changes.

I see a few paths for tweaking Clippy / Rustfmt:

1. Remove the unused fields. This is the easiest for me because then the code is gone and there's no follow up needed. (also [YAGNI](https://martinfowler.com/bliki/Yagni.html) 😉).
2. Add `#[allow(unused_tuple_struct_fields)]` in a separate PR to the tool, sync the tool subtree here, rename the lint in this PR, then do a sync back to the tool. This is a *little* silly because the tool will have an `allow` that doesn't really appear to be doing anything because nothing is warning or denying that lint to start with. There's also more steps and synchronization involved.
3. Add `#[allow(dead_code)]` in a separate PR to the tool, sync the tool subtree here, merge this PR without touching the tool. This is a *little* silly because the tool will have an `allow` that doesn't really appear to be doing anything because that code doesn't currently trigger the `dead_code` lint. If this PR is not merged for whatever reason, the tool will have a non-sensical `allow`. 

This is all based on my nebulous understanding of the tool sync process, so it's quite possible I misunderstand something.

File: src/tools/clippy/tests/ui/borrow_interior_mutable_const/auxiliary/helper.rs
Comment (shepmaster) [reply src/tools/clippy/tests/ui/borrow_interior_mutable_const/auxiliary/helper.rs:5]: Sounds good. These ui tests aren't quite the same as some of the [other, special, ui tests](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Moving.20.60unused_tuple_struct_fields.60.20from.20allow.20to.20warn/near/404297112) so I didn't think to just remove it.

**Comment by shepmaster** [comment]
Ok, my proposed set of commits for this PR is:

1. Address issues in Rustfmt
2. Address issues in Clippy
3. Combine the lints / fix the tests

**Comment by shepmaster** [comment]
I plan to address the conflicts when the review is completed; I expect that keeping up with upstream changes will be tiresome so I might as well batch them.

**Comment by shepmaster** [comment]
@WaffleLapkin  let me know what you’d like to see changed, thanks!

File: compiler/rustc_passes/src/dead.rs
Comment (WaffleLapkin) [reply compiler/rustc_passes/src/dead.rs]: > I thought you were suggesting that we should move it out of the loop — I must not understand what you mean.

Right, I see why this is confusing. 

> "trying to move _this_ out of the loop"

_this_ refers to the `is_pos` which is moved into `is_positional`, which is outside of the loop.

**Comment by WaffleLapkin** [comment]
@shepmaster I don't want to block this PR on refactoring the lint, so I think the main thing I want from you is to split the 57edc5e0139a5d6d62c42729f28af2123530ecfc so the test fixes are in a separate commit (or even PR (ideally library changes are also separate commit/PR)). (and then also squash fixups into their respective commits + fix merge conflics, obv).

Also, is it possible to fill clippy/rustfmt changes to their respective repos? I don't think we lint on unused `#[allow]`, so you can just add them maybe? Unless I'm missing something.

**Comment by WaffleLapkin** [comment]
@shepmaster what's the status here?

**Comment by shepmaster** [comment]
The current rough plan is to have the following commits in this PR:

1. fix clippy production code
1. fix rustfmt production code (add `allow`)
1. changing the compiler / renaming the lint / warn-by-default
1. fix library code
1. fix the tests (rustc)
1. fix the tests (clippy)
1. fix rustfmt production code (rename `allow`)
1. fix the tests (rustfmt)

**Comment by cjgillot** [comment]
Should we add `unused_tuple_struct_fields` to the list of renamed lints?

**Comment by shepmaster** [comment]
> Should we add `unused_tuple_struct_fields` to the list of renamed lints?

I can do that, but renamed or removed?

**Review by WaffleLapkin — APPROVED** [review]
> I can do that, but renamed or removed?

I think renamed, since it was replaced, even if by an old lint.

----

r=me with the rename

**Comment by shepmaster** [comment]
> Pretty much all of the big regressions happened for the [`issue-46449`](https://github.com/rust-lang/rustc-perf/tree/master/collector/compile-benchmarks/issue-46449?rgh-link-date=2024-01-05T08%3A56%3A33Z) benchmark

That benchmark does appear to trigger the new lint:

```
issue-46449 % cargo build
   Compiling futures v0.1.17
   Compiling issue-46649-byte-filling-slowing-down-sroa v0.1.0 (/private/tmp/rustc-perf/collector/compile-benchmarks/issue-46449)
    Finished dev [unoptimized + debuginfo] target(s) in 0.56s
```

```
issue-46449 % cargo +stage1 build
   Compiling futures v0.1.17
   Compiling issue-46649-byte-filling-slowing-down-sroa v0.1.0 (/private/tmp/rustc-perf/collector/compile-benchmarks/issue-46449)
warning: field `0` is never read
 --> src/lib.rs:6:18
  |
6 | pub struct Error(::std::io::Error);
  |            ----- ^^^^^^^^^^^^^^^^
  |            |
  |            field in this struct
  |
  = note: `#[warn(dead_code)]` on by default
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
  |
6 | pub struct Error(());
  |                  ~~

warning: `issue-46649-byte-filling-slowing-down-sroa` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.57s
```

Perhaps all the new instructions are due to diagnostics happening at all. A test of that would be to apply the patch:

```diff
diff --git a/collector/compile-benchmarks/issue-46449/src/lib.rs b/collector/compile-benchmarks/issue-46449/src/lib.rs
index f1403044..131e60ea 100644
--- a/collector/compile-benchmarks/issue-46449/src/lib.rs
+++ b/collector/compile-benchmarks/issue-46449/src/lib.rs
@@ -3,7 +3,7 @@ extern crate futures;
 use futures::{Future, Poll};
 
 const BUFFER_SIZE: usize = 1;
-pub struct Error(::std::io::Error);
+pub struct Error(#[allow(dead_code)] ::std::io::Error);
 
 struct Dummy<T>(T);
 impl<T> Future for Dummy<T> {
```

**Comment by Kobzol** [comment]
Allowing the lint for this single field indeed erases the regression. And here is our culprit:
```
< 9,297,751  ???:
  4,846,394    <fluent_syntax::parser::core::Parser<&str>>::get_pattern
  1,449,265    <fluent_bundle::resource::FluentResource>::try_new
    545,930    <fluent_syntax::parser::core::Parser<&str>>::get_placeable
    386,145    <fluent_syntax::parser::core::Parser<&str>>::get_attributes
```
It looks like diagnostic translations are lazy loaded, and since before there was no diagnostic for this benchmark, this whole thing was skipped. Now there is a diagnostic, so the translations had to be loaded, which caused the big regression.

I'll let the triager decide on this, but personally I would just mark this as triaged.

**Comment by saethlin** [comment]
> It looks like diagnostic translations are lazy loaded, and since before there was no diagnostic for this benchmark, this whole thing was skipped. Now there is a diagnostic, so the translations had to be loaded, which caused the big regression.

I looked into the other regressions, match-stress and syn. Both trip the new diagnostic and have the same sort of cachegrind diff.

**Comment by shepmaster** [comment]
Let me know if I should send a PR to rustc-perf to add `allow` for these places.

**Comment by Kobzol** [comment]
That's funny, since we just today received a similar rustc-perf PR from another rustc PR that modified unused import checking :laughing: You can read [here](https://github.com/rust-lang/rustc-perf/pull/1793) about why we don't want to change the benchmarks.

**Pull request merged**
